app/Http/Controllers/Backend/Social/AbstractTwitterController.php

```diff
@@ -0,0 +1,150 @@
+use App\Models\Status;
+use App\Models\User;
+use App\Notifications\TwitterNotSent;
+use Exception;
+use Illuminate\Support\Facades\Date;
+use Illuminate\Support\Facades\Log;
+use Laravel\Socialite\Contracts\User as SocialiteUser;
+
+abstract class AbstractTwitterController extends Controller
+{
+    /**
+     * @throws NotConnectedException
+     */
+    public static function forUser(User $user): AbstractTwitterController {
+        $sPro = $user->socialProfile;
+        if ($sPro?->twitter_id === null || $sPro?->twitter_token === null) {
+            throw new NotConnectedException();
+        }
+
+        if ($sPro->twitter_tokenSecret !== null) {
+            return new LegacyTwitterController();
+        } else {
+            return new TwitterController();
+        }
+    }
+
+    /**
+     * Function to create a user with a login-provider.
+     * If logged in, the user will have the login-provider added.
+     * If a user with corresponding login-provider already exists, it will be returned.
+     *
+     * @param SocialiteUser $socialiteUser
+     *
+     * @return User model
+     */
+    public static function getUserFromSocialite(SocialiteUser $socialiteUser): User {
+        $socialProfile = SocialLoginProfile::where('twitter_id', $socialiteUser->id)->first();
+
+        if ($socialProfile !== null) {
+            self::updateToken($socialProfile->user, $socialiteUser);
+            return $socialProfile->user;
+        }
+
+        if (auth()->check()) {
+            self::updateToken(auth()->user(), $socialiteUser);
+            return auth()->user();
+        }
+        return self::createUser($socialiteUser);
+    }
+
+    private static function createUser(SocialiteUser $socialiteUser): User {
+        $user = User::create([
+                                 'name'     => SocialController::getDisplayName($socialiteUser),
+                                 'username' => SocialController::getUniqueUsername($socialiteUser->getNickname()),
+                             ]);
+        self::updateToken($user, $socialiteUser);
+        return $user;
+    }
+
+    private static function updateToken(User $user, SocialiteUser $socialiteUser): void {
+        $user->socialProfile->update([
+                                         'twitter_id'               => $socialiteUser->id,
+                                         'twitter_token'            => $socialiteUser->token,
+                                         'twitter_refresh_token'    => $socialiteUser->refreshToken,
+                                         'twitter_token_expires_at' => Date::now()->add('second', $socialiteUser->expiresIn)
+                                     ]);
+    }
+
+
+    /**
+     * @param Status $status
+     * @param string $socialText
+     *
+     * @return int
+     * @throws NotConnectedException
+     * @throws TweetNotSendException
+     */
+    public abstract function postTweet(Status $status, string $socialText): int;
```

Comment:
`$response->id` is not an integer, see discussion in https://github.com/Traewelling/traewelling/pull/1074#discussion_r980957456 and [Twitter Docs](https://developer.twitter.com/en/docs/twitter-ids).

In the database, we use a string/varchar field for this reason, and maybe that's the best here too. Then, we'd return `$response->id_str` (in the lecacy) resp. `->data->id_str` (in the new implementation).